ytypes/string_type.go

```diff
@@ -26,6 +27,57 @@ import (
 
 // Refer to: https://tools.ietf.org/html/rfc6020#section-9.4.
 
+var (
+	// posixRegexCache stores previously-compiled Regexp objects.
+	// This helps the performance of validation of, say, a large prefix
+	// list that have the same pattern specification.
+	//
+	// Concurrency Requirements
+	//
+	// Only the regexp cache has to be protected by mutexes, since
+	// a Regexp is safe for concurrent use by multiple goroutines:
+	// https://golang.org/src/regexp/regexp.go
+	posixRegexCache = map[string]*regexp.Regexp{}
+	posixRegexMutex = &sync.RWMutex{}
+	re2RegexCache   = map[string]*regexp.Regexp{}
+	re2RegexMutex   = &sync.RWMutex{}
+)
+
+// compilePattern returns the compiled regex for the given regex
+// pattern. It caches previous look-ups for faster performance.
+// Go's regexp implementation might be relatively slow compared to other
+// languages: https://github.com/golang/go/issues/11646
+func compilePattern(pattern string, isPOSIX bool) (*regexp.Regexp, error) {
+	regexCache := re2RegexCache
+	regexMutex := re2RegexMutex
+	regexCompile := regexp.Compile
+	if isPOSIX {
+		regexCache = posixRegexCache
+		regexCompile = regexp.CompilePOSIX
+		regexMutex = posixRegexMutex
+	}
+
+	regexMutex.RLock()
+	r, ok := regexCache[pattern]
+	regexMutex.RUnlock()
+	if !ok {
+		var err error
+		r, err = regexCompile(pattern)
+		if err != nil {
+			return nil, err
+		}
+		// It's true that there may be multiple writers into the map at
+		// the same time. This, however, doesn't impact correctness,
```

Comment:
There aren't multiple writers at the same time given the mutex right? I think we mean here "It does not matter which instance of pattern is introduced into the map, since the compiled regexp.Regexp are equal. If we reach this codepath and another writer has already introduced the pattern to the map whilst we waited to acquire the lock, it will be overwritten with identical contents."

Reply:
Done with some tweaks.